rfc/rfc-51/rfc-51.md

```diff
@@ -0,0 +1,233 @@
+
+To implement this capacity, we have to upgrade the write and read parts. Let hudi can figure out the changing data when read. And in some cases, writing the extra data to help querying the changing data if necessary.
+
+## Scenario Definition
+
+Here use a simply case to explain the CDC.
+
+![](scenario-definition.jpg)
+
+Here one metadata column named `_changing_type` is added. It represents that how the record is changed, and it have four enum values:
+
+- update_preimage: The old value before a certain commit;
+- update_postiamge: The new value after a certain commit;
+- insert: a new coming record in a certain commit;
+- delete: a record that has been deleted in a certain commit;
+
+Notice:
+Here the illustration ignores all the metadata columns like `_hoodie_commit_time`.
+
+## Goals
+
+1. Support row-level CDC records generation and persistence;
+1. Support both MOR and COW tables;
+1. Support all the write operations;
+1. Support Spark DataFrame/SQL/Streaming Query;
+
+## Implementation
+### CDC Architecture
+
+![](arch.jpg)
+
+Notice:
+Other operations like `Compact`, `Clean`, `Index` do not write/change any data. So we don't need to consider in CDC scenario.
+ 
+### Config Definitions
+
+|  | default |  |
+| --- | --- | --- |
+| hoodie.table.cdf.enabled | false | if true, write the changing data to FS. |
+|  |  |  |
+| hoodie.datasource.read.cdc.enabled | false | if true, return the CDC data. |
+| hoodie.datasource.read.start.timestamp | - | requried. |
+| hoodie.datasource.read.end.timestamp | - | optional. |
+
+
+### Write
+
+Hoodie writes data by `HoodieWriteHandle`. In the different sub classes of `HoodieWriteHandle`, we will create `FileWriter`which can receive data and save to `FileSystem`. So We can upgrade these sub classes to archieve the CDC data's generation and persistence.
+
+The directory of the CDC file is`tablePath/.cdc/`. The file structure is like:
+For non-partition table:
+```
+hudi_cdc_table/
+    .hoodie/
+        hoodie.properties
+        00001.commit
+        00002.replacecommit
+        ...
+    .cdc/
+        xxxx123.parquet
+        xxxx456.parquet
+        ...
+    default/
+        fileId1_xxx_00001.parquet
+        fileId1_xxx_00002.parquet
+        ...
+```
+
+For partition table (the partition column is `year`):
+```
+hudi_cdc_table/
+    .hoodie/
+        hoodie.properties
+        00001.commit
+        00002.replacecommit
+        ...
+    .cdc/
+        year=2021/xxxx123.parquet
+        year=2022/xxxx456.parquet
+        ...
+    default/
+        year=2021/
+        year=2022/
+        ...
+```
+
+One Design Idea is that **Write CDC files as little as possible, and reuse data files as much as possible**.
```

Comment:
I am not sure I follow along here. are you suggesting this as possible design approach

Reply:
yep. After we discuss and reach an agreement, i will submit a POC asap to verify it.